File: PoolTrajectoryCleanroom/app/src/main/java/com/example/pooltrajectory/MainActivity.java
package com.example.pooltrajectory;

import android.Manifest;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Color;
import android.media.projection.MediaProjectionManager;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.provider.Settings;
import android.view.Gravity;
import android.view.View;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;

public class MainActivity extends Activity {
    private static final int REQ_CAPTURE = 501;
    private static final int REQ_NOTIFICATIONS = 502;
    private MediaProjectionManager projectionManager;
    private TextView status;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        projectionManager = (MediaProjectionManager) getSystemService(Context.MEDIA_PROJECTION_SERVICE);

        LinearLayout root = new LinearLayout(this);
        root.setOrientation(LinearLayout.VERTICAL);
        root.setGravity(Gravity.CENTER_HORIZONTAL);
        root.setPadding(dp(24), dp(40), dp(24), dp(24));
        root.setBackgroundColor(Color.rgb(245, 245, 245));

        TextView title = new TextView(this);
        title.setText("Pool Trajectory Lab");
        title.setTextSize(26f);
        title.setTextColor(Color.BLACK);
        title.setGravity(Gravity.CENTER);
        root.addView(title, matchWrap(dp(10)));

        TextView info = new TextView(this);
        info.setText("Clean-room prototype: screen capture → table validation → cue-ball/guide detection → trajectory overlay.\n\nIt refuses to draw when a pool-table scene cannot be validated.");
        info.setTextSize(16f);
        info.setTextColor(Color.DKGRAY);
        info.setGravity(Gravity.CENTER);
        root.addView(info, matchWrap(dp(28)));

        Button overlay = new Button(this);
        overlay.setText("1. Allow overlay");
        overlay.setOnClickListener(v -> requestOverlay());
        root.addView(overlay, matchWrap(dp(8)));

        Button start = new Button(this);
        start.setText("2. Start detector");
        start.setOnClickListener(v -> startDetector());
        root.addView(start, matchWrap(dp(8)));

        Button stop = new Button(this);
        stop.setText("Stop detector");
        stop.setOnClickListener(v -> {
            stopService(new Intent(this, CaptureService.class));
            setStatus("Stopped");
        });
        root.addView(stop, matchWrap(dp(18)));

        status = new TextView(this);
        status.setTextSize(15f);
        status.setTextColor(Color.BLACK);
        status.setText("Status: ready");
        status.setGravity(Gravity.CENTER);
        root.addView(status, matchWrap(0));

        setContentView(root);
        maybeAskNotificationPermission();
    }

    private LinearLayout.LayoutParams matchWrap(int bottomMargin) {
        LinearLayout.LayoutParams p = new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.WRAP_CONTENT);
        p.bottomMargin = bottomMargin;
        return p;
    }

    private int dp(int value) {
        return Math.round(value * getResources().getDisplayMetrics().density);
    }

    private void maybeAskNotificationPermission() {
        if (Build.VERSION.SDK_INT >= 33 && checkSelfPermission(Manifest.permission.POST_NOTIFICATIONS)
                != PackageManager.PERMISSION_GRANTED) {
            requestPermissions(new String[]{Manifest.permission.POST_NOTIFICATIONS}, REQ_NOTIFICATIONS);
        }
    }

    private void requestOverlay() {
        if (Settings.canDrawOverlays(this)) {
            setStatus("Overlay permission already granted");
            return;
        }
        Intent i = new Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION,
                Uri.parse("package:" + getPackageName()));
        startActivity(i);
    }

    private void startDetector() {
        if (!Settings.canDrawOverlays(this)) {
            setStatus("Grant overlay permission first");
            requestOverlay();
            return;
        }
        setStatus("Waiting for screen-capture permission…");
        startActivityForResult(projectionManager.createScreenCaptureIntent(), REQ_CAPTURE);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode != REQ_CAPTURE) return;
        if (resultCode != RESULT_OK || data == null) {
            setStatus("Screen capture denied");
            return;
        }

        Intent service = new Intent(this, CaptureService.class);
        service.putExtra(CaptureService.EXTRA_RESULT_CODE, resultCode);
        service.putExtra(CaptureService.EXTRA_RESULT_DATA, data);
        if (Build.VERSION.SDK_INT >= 26) startForegroundService(service); else startService(service);
        setStatus("Detector started — open the pool game");
    }

    private void setStatus(String message) {
        if (status != null) status.setText("Status: " + message);
    }
}

File: app/src/main/java/com/example/trajectoryoverlay/MainActivity.kt
package com.example.trajectoryoverlay

import android.Manifest
import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Color
import android.media.projection.MediaProjectionManager
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.provider.Settings
import android.view.Gravity
import android.view.ViewGroup
import android.widget.*
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import org.opencv.android.OpenCVLoader

class MainActivity : AppCompatActivity() {
    private lateinit var status: TextView
    private lateinit var projectionManager: MediaProjectionManager
    private var openCvReady = false

    private val captureLauncher = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
        if (result.resultCode == Activity.RESULT_OK && result.data != null) {
            val i = Intent(this, CaptureOverlayService::class.java).apply {
                action = CaptureOverlayService.ACTION_START
                putExtra(CaptureOverlayService.EXTRA_RESULT_CODE, result.resultCode)
                putExtra(CaptureOverlayService.EXTRA_RESULT_DATA, result.data)
            }
            ContextCompat.startForegroundService(this, i)
            status.text = "Captura autorizada. Abra a mesa e veja o diagnóstico no topo do overlay."
        } else {
            status.text = "Captura de tela não autorizada."
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        projectionManager = getSystemService(MediaProjectionManager::class.java)
        Prefs.applyV07Defaults(this)
        openCvReady = OpenCVLoader.initLocal()
        requestNotificationPermission()
        setContentView(buildUi())
    }

    override fun onResume() {
        super.onResume()
        if (::status.isInitialized) updateLocalStatus()
    }

    private fun buildUi(): ScrollView {
        val scroll = ScrollView(this)
        scroll.setBackgroundColor(Color.rgb(11, 15, 18))
        val root = LinearLayout(this).apply {
            orientation = LinearLayout.VERTICAL
            setPadding(dp(22), dp(18), dp(22), dp(18))
        }
        scroll.addView(root)

        root.addView(TextView(this).apply {
            text = "Trajectory Overlay Lab v0.7"
            textSize = 25f
            setTextColor(Color.WHITE)
        })
        root.addView(TextView(this).apply {
            text = "v0.7: detector de bolas muito mais restrito. Só aceita círculos com tamanho de bola real, dentro do pano e cercados pelo feltro; círculos falsos de HUD/decoração são rejeitados."
            textSize = 14f
            setTextColor(Color.rgb(185, 199, 210))
            setPadding(0, dp(6), 0, dp(14))
        })

        val row1 = LinearLayout(this).apply { orientation = LinearLayout.HORIZONTAL }
        row1.addView(button("1. Permitir overlay") { openOverlayPermission() })
        row1.addView(button("2. Testar overlay") { testOverlay() })
        root.addView(row1)

        val row2 = LinearLayout(this).apply { orientation = LinearLayout.HORIZONTAL }
        row2.addView(button("3. Iniciar análise") { startCapture() })
        row2.addView(button("Parar") { stopCapture() })
        root.addView(row2)

        status = TextView(this).apply {
            textSize = 13f
            setTextColor(Color.rgb(124, 203, 255))
            setPadding(0, dp(12), 0, dp(12))
        }
        root.addView(status)
        updateLocalStatus()

        root.addView(check("Trajetórias diretas", Prefs.showDirect(this)) { Prefs.setDirect(this, it) })
        root.addView(check("Bank shots (1 tabela)", Prefs.showBanks(this)) { Prefs.setBanks(this, it) })
        root.addView(check("Trajetória da branca após a colisão", Prefs.showSecondary(this)) { Prefs.setSecondary(this, it) })
        root.addView(check("Mostrar várias opções ao mesmo tempo", Prefs.showAll(this)) { Prefs.setAll(this, it) })
        root.addView(check("Mostrar diagnóstico visual (mesa/bolas)", Prefs.visualDebug(this)) { Prefs.setVisualDebug(this, it) })

        root.addView(TextView(this).apply {
            text = "Sensibilidade de detecção das bolas"
            setTextColor(Color.WHITE)
            setPadding(0, dp(12), 0, 0)
        })
        val sensText = TextView(this).apply {
            setTextColor(Color.LTGRAY)
            text = "${Prefs.sensitivity(this@MainActivity)} (menor = detecta mais)"
        }
        val seek = SeekBar(this).apply {
            max = 20
            progress = (Prefs.sensitivity(this@MainActivity) - 10).coerceIn(0, 20)
            setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
                override fun onProgressChanged(s: SeekBar?, progress: Int, fromUser: Boolean) {
                    val value = 10 + progress
                    Prefs.setSensitivity(this@MainActivity, value)
                    sensText.text = "$value (menor = detecta mais)"
                }
                override fun onStartTrackingTouch(s: SeekBar?) {}
                override fun onStopTrackingTouch(s: SeekBar?) {}
            })
        }
        root.addView(seek)
        root.addView(sensText)

        root.addView(TextView(this).apply {
            text = "Diagnóstico esperado\n• Teste: aparece um X azul, texto e linhas de exemplo por 8 s.\n• Durante a análise, puxe a notificação Trajectory Overlay e toque em ‘Salvar print’.\n• O app salva na Galeria/Pictures/TrajectoryOverlay uma imagem da tela capturada com mesa, bolas e trajetórias marcadas.\n• Envie essa imagem para eu calibrar o detector.\n• Deixe ‘Mostrar diagnóstico visual’ desligado para a tela ficar limpa."
            textSize = 13f
            setTextColor(Color.rgb(190, 200, 208))
            setPadding(0, dp(18), 0, dp(8))
        })
        return scroll
    }

    private fun updateLocalStatus() {
        val overlayOk = Settings.canDrawOverlays(this)
        status.text = "Overlay: ${if (overlayOk) "OK" else "SEM PERMISSÃO"} • OpenCV: ${if (openCvReady) "OK" else "FALHOU"}"
    }

    private fun button(text: String, onClick: () -> Unit) = Button(this).apply {
        this.text = text
        setOnClickListener { onClick() }
        layoutParams = LinearLayout.LayoutParams(0, ViewGroup.LayoutParams.WRAP_CONTENT, 1f).apply { marginEnd = dp(8) }
    }

    private fun check(text: String, checked: Boolean, changed: (Boolean) -> Unit) = CheckBox(this).apply {
        this.text = text
        isChecked = checked
        setTextColor(Color.WHITE)
        textSize = 14f
        gravity = Gravity.CENTER_VERTICAL
        setOnCheckedChangeListener { _, v -> changed(v) }
    }

    private fun openOverlayPermission() {
        startActivity(Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION, Uri.parse("package:$packageName")))
    }

    private fun testOverlay() {
        if (!Settings.canDrawOverlays(this)) {
            status.text = "Conceda a permissão de overlay primeiro."
            openOverlayPermission()
            return
        }
        startService(Intent(this, CaptureOverlayService::class.java).apply {
            action = CaptureOverlayService.ACTION_TEST
        })
        status.text = "Teste iniciado por 8 segundos. Saia do app e veja se o X/linhas aparecem."
    }

    private fun startCapture() {
        if (!Settings.canDrawOverlays(this)) {
            status.text = "Conceda a permissão de overlay primeiro."
            openOverlayPermission()
            return
        }
        if (!openCvReady) {
            status.text = "OpenCV não carregou; reinstale a versão mais recente do APK."
            return
        }
        captureLauncher.launch(projectionManager.createScreenCaptureIntent())
    }

    private fun stopCapture() {
        startService(Intent(this, CaptureOverlayService::class.java).apply { action = CaptureOverlayService.ACTION_STOP })
        status.text = "Análise parada."
    }

    private fun requestNotificationPermission() {
        if (Build.VERSION.SDK_INT >= 33 && checkSelfPermission(Manifest.permission.POST_NOTIFICATIONS) != PackageManager.PERMISSION_GRANTED) {
            requestPermissions(arrayOf(Manifest.permission.POST_NOTIFICATIONS), 10)
        }
    }

    private fun dp(v: Int) = (v * resources.displayMetrics.density).toInt()
}
